### PR TITLE
Refactor AMP Param Parsing

### PR DESCRIPTION
**Pull request by SyntaxNode**
We're starting to separate amp specific functionality into it's own package, with the goal of creating slim endpoints funcs. Parsing is indirectly tested by pre-existing AMP tests. I'll need to add proper coverage before converting from a draft.

### CODE REVIEW

**Comment by SyntaxNode** [comment]
Removing stale label to end of year holidays.

**Review by hhhjort — APPROVED** [review]
LGTM

File: endpoints/openrtb2/amp_auction.go [hunk 1]
```diff
@@ -406,57 +399,52 @@ func (deps *endpointDeps) overrideWithParams(httpRequest *http.Request, req *ope
...
 		return []error{policyWriterErr}
 	}
 	if err := policyWriter.Write(req); err != nil {
 		return []error{err}
 	}
 
-	if timeout, err := strconv.ParseInt(httpRequest.FormValue("timeout"), 10, 64); err == nil {
-		req.TMax = timeout - deps.cfg.AMPTimeoutAdjustment
+	if ampParams.Timeout != nil {
+		req.TMax = int64(*ampParams.Timeout) - deps.cfg.AMPTimeoutAdjustment
```
Comment (guscarreon) [inline endpoints/openrtb2/amp_auction.go:417]: Is there a possibility that `req.TMax` ends up being negative? Should we put a check?

File: amp/parse.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+		Size: Size{
+			Height:         parseInt(query.Get("h")),
+			Multisize:      parseMultisize(query.Get("ms")),
+			OverrideHeight: parseInt(query.Get("oh")),
+			OverrideWidth:  parseInt(query.Get("ow")),
+			Width:          parseInt(query.Get("w")),
+		},
+		Slot:            query.Get("slot"),
+		StoredRequestID: tagID,
+		Timeout:         parseIntPtr(query.Get("timeout")),
```
Comment (guscarreon) [inline amp/parse.go:58]: Why do we need it to be a pointer? Not saying it is wrong, I just noticed it wasn't a pointer before.

File: endpoints/openrtb2/amp_auction.go (hunk 1 above)
Comment (SyntaxNode) [reply endpoints/openrtb2/amp_auction.go:417]: I'm not sure. Could you please create an internal ticket for us to revisit?

File: endpoints/openrtb2/amp_auction.go [hunk 3]
```diff
@@ -406,57 +399,52 @@ func (deps *endpointDeps) overrideWithParams(httpRequest *http.Request, req *ope
...
 	}
 
 	return nil
 }
 
-func makeFormatReplacement(overrideWidth uint64, overrideHeight uint64, width uint64, height uint64, multisize string) []openrtb.Format {
+func makeFormatReplacement(size amp.Size) []openrtb.Format {
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply endpoints/openrtb2/amp_auction.go:423]: Possibly. This is strictly a refactor PR. You're welcomed to explore that logic change in a separate PR.

File: amp/parse.go (hunk 2 above)
Comment (SyntaxNode) [reply amp/parse.go:58]: I'm using a pointer to distinguish if a timeout value was provided and valid versus not provided and/or invalid. Previously,  the parse method was in-line with its usage and we checked for a parse error immediately. Now there's the need to communicate a parse error differently. Go doesn't have many constructs for doing so, sadly.

File: endpoints/openrtb2/amp_auction.go [hunk 4]
```diff
@@ -367,35 +365,30 @@ func (deps *endpointDeps) loadRequestJSONForAmp(httpRequest *http.Request) (req
...
 			req.Imp[0].Banner.Format = format
-		} else if width != 0 {
-			setWidths(req.Imp[0].Banner.Format, width)
-		} else if height != 0 {
-			setHeights(req.Imp[0].Banner.Format, height)
+		} else if ampParams.Size.Width != 0 {
+			setWidths(req.Imp[0].Banner.Format, ampParams.Size.Width)
+		} else if ampParams.Size.Height != 0 {
+			setHeights(req.Imp[0].Banner.Format, ampParams.Size.Height)
 		}
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply endpoints/openrtb2/amp_auction.go:385]: Possibly. With this PR I'm refactoring the parse flow, not the implementation.

**Review by guscarreon — APPROVED** [review]
Created internal ticket to list come of the remaining comments. LGTM

**Pull request merged**